Detectors/ZDC/simulation/src/Digits2Raw.cxx

```diff
@@ -0,0 +1,874 @@
+void Digits2Raw::setTriggerMask()
+{
+  mTriggerMask = 0;
+  mPrintTriggerMask = "";
+  for (Int_t im = 0; im < NModules; im++) {
+    if (im > 0)
+      mPrintTriggerMask += " ";
+    mPrintTriggerMask += std::to_string(im);
+    mPrintTriggerMask += "[";
+    for (UInt_t ic = 0; ic < NChPerModule; ic++) {
+      if (mModuleConfig->modules[im].trigChannel[ic]) {
+        UInt_t tmask = 0x1 << (im * NChPerModule + ic);
+        mTriggerMask = mTriggerMask | tmask;
+        mPrintTriggerMask += "T";
+      } else {
+        mPrintTriggerMask += " ";
+      }
+    }
+    mPrintTriggerMask += "]";
+    UInt_t mytmask = mTriggerMask >> (im * NChPerModule);
+    printf("Trigger mask for module %d 0123 %s%s%s%s\n", im,
+           mytmask & 0x1 ? "T" : "N",
+           mytmask & 0x2 ? "T" : "N",
+           mytmask & 0x4 ? "T" : "N",
+           mytmask & 0x8 ? "T" : "N");
+  }
+  printf("trigger_mask=0x%08x %s\n", mTriggerMask, mPrintTriggerMask.data());
+}
+
+//______________________________________________________________________________
+inline void Digits2Raw::resetSums(uint32_t orbit)
+{
+  for (Int_t im = 0; im < NModules; im++) {
+    for (Int_t ic = 0; ic < NChPerModule; ic++) {
+      mScalers[im][ic] = 0;
+      mSumPed[im][ic] = 0;
+      mPed[im][ic] = 0;
+    }
+  }
+  mLastOrbit = orbit;
+  mLastNEmpty = 0;
+}
+
+//______________________________________________________________________________
+inline void Digits2Raw::updatePedestalReference(int bc)
+{
+  // Compute or update baseline reference
+  if (mEmpty[bc] > 0 && mEmpty[bc] != mLastNEmpty) {
+    for (Int_t im = 0; im < NModules; im++) {
+      for (Int_t ic = 0; ic < NChPerModule; ic++) {
+        // Identify connected channel
+        auto id = mModuleConfig->modules[im].channelID[ic];
+        auto base_m = mSimCondition->channels[id].pedestal;      // Average pedestal
+        auto base_s = mSimCondition->channels[id].pedestalFluct; // Baseline oscillations
+        auto base_n = mSimCondition->channels[id].pedestalNoise; // Electronic noise
+        Double_t deltan = mEmpty[bc] - mLastNEmpty;
+        // We assume to have a fluctuation every two bunch crossings
+        // Will need to tune this parameter
+        Double_t k = 2.;
+        mSumPed[im][ic] += gRandom->Gaus(12. * deltan * base_m, 12. * k * base_s * TMath::Sqrt(deltan / k));
+        // Adding in quadrature the RMS of pedestal electronic noise
+        mSumPed[im][ic] += gRandom->Gaus(0, base_n * TMath::Sqrt(12. * deltan));
+        Double_t myped = TMath::Nint(8. * mSumPed[im][ic] / Double_t(mEmpty[bc]) / 12. + 32768);
+        if (myped < 0)
+          myped = 0;
+        if (myped > 65535)
+          myped = 65535;
+        mPed[im][ic] = myped;
+      }
+    }
+    mLastNEmpty = mEmpty[bc];
+  }
+}
+
+//______________________________________________________________________________
+void Digits2Raw::insertLastBunch(int ibc, uint32_t orbit)
```

Comment:
Cannot this method be implemented as a custom case of convertDigits (e.g. with some extra option). Most of the code is duplicated, this is quite error prone.

Reply:
Good suggestion. I moved the duplicated code to private methods of the class. In this way the differences in behavior are more clear.